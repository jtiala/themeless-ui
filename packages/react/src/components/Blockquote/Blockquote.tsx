import { CommonComponentProps, cn } from "@themeless-ui/utils";
import { ReactNode } from "react";

export type BlockquoteProps = {
  /**
   * A URI of the source of the information quoted.
   */
  cite?: string;

  /**
   * Author of the quotation.
   */
  author?: string;

  /**
   * Source of the quotation.
   */
  source?: string;

  /**
   * Quotation content.
   */
  children?: ReactNode;
} & CommonComponentProps;

const blockquoteClassName = cn("blockquote");
const quoteClassName = cn("blockquote-quote");
const authorClassName = cn("blockquote-author");
const sourceClassName = cn("blockquote-source");

/**
 * Display an extended quotation.
 */
export function Blockquote({
  cite,
  author,
  source,
  children,
  id,
  testId,
}: BlockquoteProps) {
  return (
    <blockquote
      cite={cite}
      className={blockquoteClassName}
      id={id}
      data-testid={testId}
    >
      <div className={quoteClassName}>{children}</div>
      {(author || source) && (
        <footer>
          {author && <span className={authorClassName}>&mdash;{author}</span>}
          {author && source && <span>, </span>}
          {source && (
            <cite key="source" className={sourceClassName}>
              {source}
            </cite>
          )}
        </footer>
      )}
    </blockquote>
  );
}
