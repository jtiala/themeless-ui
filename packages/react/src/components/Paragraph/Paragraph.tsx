import { CommonComponentProps, cn } from "@themeless-ui/utils";
import { ReactNode } from "react";

export type ParagraphProps = {
  /**
   * Paragraph content.
   */
  children?: ReactNode;
} & CommonComponentProps;

const className = cn("paragraph");

/**
 * A paragraph of text.
 */
export function Paragraph({ children, id, testId }: ParagraphProps) {
  return (
    <p className={className} id={id} data-testid={testId}>
      {children}
    </p>
  );
}
