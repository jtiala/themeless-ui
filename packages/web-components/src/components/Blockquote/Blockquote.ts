import style from "@themeless-ui/style/dist/blockquote.css?inline";
import { cn } from "@themeless-ui/utils";
import { html, unsafeCSS } from "lit";
import { customElement, property } from "lit/decorators.js";
import { TUIComponent } from "../../utils";

const blockquoteClassName = cn("blockquote");
const quoteClassName = cn("blockquote-quote");
const authorClassName = cn("blockquote-author");
const sourceClassName = cn("blockquote-source");

@customElement("tui-blockquote")
export class Blockquote extends TUIComponent {
  /**
   * A URI of the source of the information quoted.
   */
  @property()
  cite?: string;

  /**
   * Author of the quotation.
   */
  @property()
  author?: string;

  /**
   * Source of the quotation.
   */
  @property()
  source?: string;

  static styles = unsafeCSS(style);

  override render() {
    const authorElement = this.author
      ? html`<span className="${authorClassName}">&mdash;${this.author}</span>`
      : undefined;

    const sourceElement = this.source
      ? html`<cite key="source" className="${sourceClassName}">
          ${this.source}
        </cite>`
      : undefined;

    const divider =
      this.author && this.source ? html`<span>, </span>` : undefined;

    const footer =
      authorElement || sourceElement
        ? html`<footer>${authorElement}${divider}${sourceElement}</footer>`
        : undefined;

    return html`
      <blockquote
        cite="${this.cite}"
        class="${blockquoteClassName}"
        id="${this.id}"
        data-testid="${this.testId}"
      >
        <div className="${quoteClassName}"><slot></slot></div>
        ${footer}
      </blockquote>
    `;
  }
}

declare global {
  interface HTMLElementTagNameMap {
    "tui-blockquote": Blockquote;
  }
}
