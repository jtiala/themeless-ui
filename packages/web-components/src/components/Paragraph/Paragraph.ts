import style from "@themeless-ui/style/dist/paragraph.css?inline";
import { cn } from "@themeless-ui/utils";
import { html, unsafeCSS } from "lit";
import { customElement } from "lit/decorators.js";
import { TUIComponent } from "../../utils";

const className = cn("paragraph");

@customElement("tui-paragraph")
export class Paragraph extends TUIComponent {
  static styles = unsafeCSS(style);

  override render() {
    return html`
      <p class="${className}" id="${this.id}" data-testid="${this.testId}">
        <slot></slot>
      </p>
    `;
  }
}

declare global {
  interface HTMLElementTagNameMap {
    "tui-paragraph": Paragraph;
  }
}
