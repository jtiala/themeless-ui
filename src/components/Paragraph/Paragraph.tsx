import { ReactNode } from "react";
import { cn } from "../../utils/css";

export type ParagraphProps = {
  /**
   * Paragraph content.
   */
  children?: ReactNode;
};

const className = cn("paragraph");

/**
 * Display a text paragraph.
 */
export function Paragraph({ children }: ParagraphProps) {
  return <p className={className}>{children}</p>;
}
