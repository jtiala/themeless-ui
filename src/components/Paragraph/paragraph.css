.tui__paragraph {
  display: block;
  margin: 0 0 var(--tui__spacing-8) 0;
  color: var(--tui__color-black);
}
