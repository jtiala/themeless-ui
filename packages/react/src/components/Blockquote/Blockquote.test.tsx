import {
  componentToJson,
  describe,
  expect,
  it,
  render,
  renderer,
  screen,
} from "../../../test";
import { Paragraph } from "../Paragraph";
import { Blockquote } from "./Blockquote";

const cite = "https://source-of-the-quote.example.com";
const author = "Samuel L. Ipsum";
const source = "Lorem Ipsum";
const lipsum =
  "Lorem ipsum dolor sit amet, consectetur adipiscing elit. Donec accumsan ex vel lacinia volutpat. Ut ultricies, mauris et varius finibus, nulla ante efficitur libero, eget convallis odio ex ut neque. Praesent quis odio ac felis pretium pharetra sit amet quis neque. Nam nec libero purus. Maecenas sagittis lobortis lacinia. Aliquam non vehicula metus. Vivamus fringilla ante eget leo pretium cursus. Morbi venenatis diam metus, ut faucibus dolor mollis a. Nulla nec ex sapien. Curabitur a tempor est. Mauris faucibus congue mi, sit amet pulvinar justo porttitor nec. Nunc et eros a est fermentum lacinia id at nisi.";
const children = <Paragraph>{lipsum}</Paragraph>;

const defaultBlockquote = (
  <Blockquote testId="blockquote" cite={cite} author={author} source={source}>
    {children}
  </Blockquote>
);

const blockquoteWithoutSource = (
  <Blockquote testId="blockquote" cite={cite} author={author}>
    {children}
  </Blockquote>
);

const blockquoteWithoutAuthor = (
  <Blockquote testId="blockquote" cite={cite} source={source}>
    {children}
  </Blockquote>
);

describe("Blockquote", async () => {
  it("should render default blockquote", () => {
    expect(
      componentToJson(renderer.create(defaultBlockquote)),
    ).toMatchSnapshot();

    render(defaultBlockquote);

    const blockquote = screen.getByTestId("blockquote");
    const footer = blockquote.querySelector("footer");

    expect(blockquote).toBeInTheDocument();
    expect(blockquote).toHaveTextContent(lipsum);
    expect(blockquote).toHaveAttribute("cite", cite);
    expect(footer?.textContent).toBe(`—${author}, ${source}`);
  });

  it("should render author without source", () => {
    expect(
      componentToJson(renderer.create(blockquoteWithoutSource)),
    ).toMatchSnapshot();

    render(blockquoteWithoutSource);

    const blockquote = screen.getByTestId("blockquote");
    const footer = blockquote.querySelector("footer");

    expect(footer?.textContent).toBe(`—${author}`);
  });

  it("should render source without author", () => {
    expect(
      componentToJson(renderer.create(blockquoteWithoutAuthor)),
    ).toMatchSnapshot();

    render(blockquoteWithoutAuthor);

    const blockquote = screen.getByTestId("blockquote");
    const footer = blockquote.querySelector("footer");

    expect(footer?.textContent).toBe(source);
  });
});
