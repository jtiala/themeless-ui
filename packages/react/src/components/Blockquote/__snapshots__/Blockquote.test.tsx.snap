// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Blockquote > should render author without source 1`] = `
<blockquote
  cite="https://source-of-the-quote.example.com"
  className="tui__blockquote"
  data-testid="blockquote"
>
  <div
    className="tui__blockquote-quote"
  >
    <p
      className="tui__paragraph"
    >
      Lorem ipsum dolor sit amet, consectetur adipiscing elit. Donec accumsan ex vel lacinia volutpat. Ut ultricies, mauris et varius finibus, nulla ante efficitur libero, eget convallis odio ex ut neque. Praesent quis odio ac felis pretium pharetra sit amet quis neque. Nam nec libero purus. Maecenas sagittis lobortis lacinia. Aliquam non vehicula metus. Vivamus fringilla ante eget leo pretium cursus. Morbi venenatis diam metus, ut faucibus dolor mollis a. Nulla nec ex sapien. Curabitur a tempor est. Mauris faucibus congue mi, sit amet pulvinar justo porttitor nec. Nunc et eros a est fermentum lacinia id at nisi.
    </p>
  </div>
  <footer>
    <span
      className="tui__blockquote-author"
    >
      —
      Samuel L. Ipsum
    </span>
  </footer>
</blockquote>
`;

exports[`Blockquote > should render default blockquote 1`] = `
<blockquote
  cite="https://source-of-the-quote.example.com"
  className="tui__blockquote"
  data-testid="blockquote"
>
  <div
    className="tui__blockquote-quote"
  >
    <p
      className="tui__paragraph"
    >
      Lorem ipsum dolor sit amet, consectetur adipiscing elit. Donec accumsan ex vel lacinia volutpat. Ut ultricies, mauris et varius finibus, nulla ante efficitur libero, eget convallis odio ex ut neque. Praesent quis odio ac felis pretium pharetra sit amet quis neque. Nam nec libero purus. Maecenas sagittis lobortis lacinia. Aliquam non vehicula metus. Vivamus fringilla ante eget leo pretium cursus. Morbi venenatis diam metus, ut faucibus dolor mollis a. Nulla nec ex sapien. Curabitur a tempor est. Mauris faucibus congue mi, sit amet pulvinar justo porttitor nec. Nunc et eros a est fermentum lacinia id at nisi.
    </p>
  </div>
  <footer>
    <span
      className="tui__blockquote-author"
    >
      —
      Samuel L. Ipsum
    </span>
    <span>
      ,
    </span>
    <cite
      className="tui__blockquote-source"
    >
      Lorem Ipsum
    </cite>
  </footer>
</blockquote>
`;

exports[`Blockquote > should render source without author 1`] = `
<blockquote
  cite="https://source-of-the-quote.example.com"
  className="tui__blockquote"
  data-testid="blockquote"
>
  <div
    className="tui__blockquote-quote"
  >
    <p
      className="tui__paragraph"
    >
      Lorem ipsum dolor sit amet, consectetur adipiscing elit. Donec accumsan ex vel lacinia volutpat. Ut ultricies, mauris et varius finibus, nulla ante efficitur libero, eget convallis odio ex ut neque. Praesent quis odio ac felis pretium pharetra sit amet quis neque. Nam nec libero purus. Maecenas sagittis lobortis lacinia. Aliquam non vehicula metus. Vivamus fringilla ante eget leo pretium cursus. Morbi venenatis diam metus, ut faucibus dolor mollis a. Nulla nec ex sapien. Curabitur a tempor est. Mauris faucibus congue mi, sit amet pulvinar justo porttitor nec. Nunc et eros a est fermentum lacinia id at nisi.
    </p>
  </div>
  <footer>
    <cite
      className="tui__blockquote-source"
    >
      Lorem Ipsum
    </cite>
  </footer>
</blockquote>
`;
